### ISSUE TITLE
Add decorators to specify when a REST endpoint was added

### ISSUE THREAD

**Issue by RussTheAerialist** [tableau/server-client-python#55]
Needed to be able to show users meaningful errors if they are using a version of the REST api that does not support a certain feature.  This is needed to fully address #54

**Comment by t8y8** [comment]
Currently there is only one serializer for each request type -- since the schemas aren't always compatible we'll need to deal with that somehow.

**Comment by RussTheAerialist** [comment]
I don't follow what you mean here?  As long as the major version number hasn't changed, the schemas should be compatible because we guarantee no breaking changes unless the major version is revved.  I would imagine we'll branch and no longer add features to the 2.x version once 3.x comes in.

**Comment by t8y8** [comment]
We've added attributes over time, so 2.3 may have an attribute that 2.2 does not have. If the serializer doesn't know about those differences it'll get stripped out either on read or write.

The endpoint itself is backwards compatible because we ignore attributes we don't understand and treat them all as optional.

**Comment by RussTheAerialist** [comment]
i still don't see where the problem is? If we ignore the extra attributes,
then it doesn't matter as long as the serializer includes all of them, the
rest endpoint should handle it properly.

If we changed the meaning of attributes, then that's a breaking change and
we should have revved the major version.

**Comment by t8y8** [comment]
Yeah, I guess I'm saying I don't know -- you're probably right.

I'm not aware of any places we changed the meaning of an attribute, just their existence.

**Comment by RussTheAerialist** [comment]
yeah, I think there will be some exploration that happens when we implement this enhancement.  If there are attributes we need to not serialize, we could have another decorator that says "remove this attribute if less than version x" or we could have separate serializers for each version where we need them.  It's something to figure out, but I _think_ it should all work.

**Comment by t8y8** [comment]
I hacked up a quick decorator and tested that this does in fact work:

```python
def min_api(min_ver):
    def outter_api_decorator(func):
        @wraps(func)
        def wrapper(self, *args):
            server_version = self.parent_srv.version
            if float(server_version) < min_ver:
                error = "This endpoint is not support in version {}. Requires {}".format(server_version, min_ver)
                raise ValueError(error)
            return func(self, *args)
        return wrapper
    return outter_api_decorator
```

It should probably throw a specific exception `EndpointNotSupported` or something but this was copy/paste from an example to start.

1. Where to put it? What exception to use? Will this work? Etc.
2. I think it really needs the 'auto-detect' part of the version detection function to be the most useful

/cc @RussTheAerialist

**Comment by t8y8** [comment]
Alright, I'm working on the second PR, and as predicted all the tests fail because our mocked server objects don't set a 'version' by default. Which means that technically, all of our tests have been testing against invalid URIs -- after I fix these up that would be fixed.

But it's a lot of manual effort :(

**Comment by Talvalin** [comment]
You can't say "funniest bug in our code ever" and then just leave us all hanging....

**Comment by t8y8** [comment]
```
-        self.version = 2.3
+        self.version = "2.3"
```

We were defaulting our version number to a float, `2.3` which worked because when building the url we use `{}`.format which converts to a string. But for the purposes of checking the version, it breaks :)

So, we were using the wrong type for the default version but it basically didn't matter

**Comment by RussTheAerialist** [comment]
@t8y8 I've been thinking about this a bunch.  Couldn't we just take the argument in the decorator and run it through `str()` that was the float representation would work also?  I think it would make it more user-friendly.

**Comment by RussTheAerialist** [comment]
I believe this has been addressed in v0.4.x
